### PR TITLE
Deploy at end

### PR DESCRIPTION
**Pull request by atanasenko**

### CODE REVIEW

File: takari-lifecycle-plugin-its/src/test/java/io/tesla/maven/plugins/test/MultimoduleSkipInstallDeployTest.java [hunk 1]
```diff
@@ -32,8 +32,8 @@ public void testBasic() throws Exception {
     Assert.assertTrue(new File(group, "modulea/1.0/modulea-1.0.pom").canRead());
     Assert.assertTrue(new File(group, "modulea/1.0/modulea-1.0.jar").canRead());
 
-    Assert.assertFalse(new File(group, "moduleb/1.0/modulea-1.0.pom").canRead());
-    Assert.assertFalse(new File(group, "moduleb/1.0/modulea-1.0.jar").canRead());
+    Assert.assertFalse(new File(group, "moduleb/1.0/moduleb-1.0.pom").canRead());
```
Comment (ifedorenko) [inline takari-lifecycle-plugin-its/src/test/java/io/tesla/maven/plugins/test/MultimoduleSkipInstallDeployTest.java]: probably a good idea to submit this as a separate commit

File: takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+public class DeployParticipant extends AbstractMavenLifecycleParticipant implements LogEnabled {
+
+  private static final String CONTEXT_KEY = "deploy-at-end-requests";
+
+  protected RepositorySystem repoSystem;
+  private Logger logger;
+
+  @Override
+  public void enableLogging(Logger logger) {
```
Comment (ifedorenko) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java]: please use slf4j directly

File: takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.maven.execution.MavenSession;
+import org.apache.maven.plugin.descriptor.PluginDescriptor;
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.logging.LogEnabled;
+import org.codehaus.plexus.logging.Logger;
+import org.eclipse.aether.RepositorySystem;
+import org.eclipse.aether.deployment.DeployRequest;
+import org.eclipse.aether.deployment.DeploymentException;
+
+public class DeployParticipant extends AbstractMavenLifecycleParticipant implements LogEnabled {
```
Comment (ifedorenko) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java]: I think it is better to explicitly mark this as `@Named @Singleton`.

File: takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    List<DeployRequest> deployAtEndRequests = getDeployAtEndRequests(session);
+    boolean noerrors = session.getResult().getExceptions().isEmpty();
+
+    if (!deployAtEndRequests.isEmpty() && noerrors) {
+      logger.info("");
+      logger.info("------------------------------------------------------------------------");
+      logger.info("-- Performing deploy at end                                           --");
+      logger.info("------------------------------------------------------------------------");
+
+      synchronized (deployAtEndRequests) {
```
Comment (ifedorenko) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java]: can you double-check synchronization? to me it looks either redundant (the collection is synchronizedList) or wrong (writers are not synchronized on the list). 

Personally, I would make the collection final instance variable and explicitly synchronize both readers and writers on the collection. This way correct synchronization will be much easier to assess.

File: takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      logger.info("------------------------------------------------------------------------");
+    }
+  }
+
+  public static void deployAtEnd(MavenSession session, DeployRequest deployRequest) {
+    getDeployAtEndRequests(session).add(deployRequest);
+  }
+
+  public static List<DeployRequest> getDeployAtEndRequests(MavenSession session) {
+    MavenProject topLevelProject = session.getTopLevelProject();
```
Comment (ifedorenko) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java]: I think it is better to keep deployment request list in an instance variable. It will make the code simpler to understand and maintain. It will also work better in the unlikely event when the same reactor has multiple versions of takari-lifecycle.

File: takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/Deploy.java [hunk 6]
```diff
@@ -82,10 +93,15 @@ private void installProject(MavenProject project) throws MojoExecutionException
...
-    } catch (DeploymentException e) {
-      throw new MojoExecutionException(e.getMessage(), e);
+    if (!deployAtEnd) {
+      try {
+        repositorySystem.deploy(repositorySystemSession, deployRequest);
+      } catch (DeploymentException e) {
+        throw new MojoExecutionException(e.getMessage(), e);
+      }
+    } else {
+      DeployParticipant.deployAtEnd(mavenSession, deployRequest);
```
Comment (ifedorenko) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/Deploy.java]: I think it is cleaner to "@Inject DeployParticipant deployParticipant` and use deployParticipant instance instead of static.

Comment (atanasenko) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/Deploy.java]: That was the first thing I tried (participant was @Named @Singleton), but for some reason it was a singleton per-project, and a completely differrent instance at the point of lifecycleParticipant execution.
As far as I understood, it is due to projects and the overall build having separate classRealms and each had a different instance of that participant.

File: takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  @Inject
+  public DeployParticipant(RepositorySystem repoSystem) {
+    this.repoSystem = repoSystem;
+  }
+
+  @Override
+  public void afterSessionEnd(MavenSession session) throws MavenExecutionException {
+    boolean noerrors = session.getResult().getExceptions().isEmpty();
+
+    if (!deployAtEndRequests.isEmpty() && noerrors) {
```
Comment (ifedorenko) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java]: maybe warn there was no deploy-at-the-end if noerrors==false?

File: takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+          }
+        }
+        deployAtEndRequests.clear();
+      }
+
+      log.info("------------------------------------------------------------------------");
+    }
+  }
+
+  public List<DeployRequest> getDeployAtEndRequests() {
```
Comment (ifedorenko) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java]: I'd make it package protected and comment it's for test purposes only, to make it clear this is not meant for external use

Comment (atanasenko) [inline takari-lifecycle-plugin/src/main/java/io/takari/maven/plugins/DeployParticipant.java]: Would need to move it to the same package as the test. Ok with *.install_deploy?
